src/main/java/Duke/Storage.java

```diff
@@ -0,0 +1,199 @@
+     */
+    public void saveTasksToFile(TaskList tasks) throws IOException {
+        Path filePath = Paths.get(this.filePath);
+
+        // Create the parent directory if it doesn't exist
+        if (!Files.exists(filePath.getParent())) {
+            Files.createDirectories(filePath.getParent());
+        }
+
+        List<String> encodedTasks = TaskListEncoder.encode(tasks.getTasks());
+        // Write tasks to the file
+        Files.write(filePath, encodedTasks);
+    }
+
+    /**
+     * Loads tasks from the data file.
+     *
+     * @return The loaded TaskList.
+     * @throws IOException              If an error occurs while reading the file.
+     * @throws DukeDataCorruptedException If the data file is corrupted.
+     */
+    public TaskList loadTasksFromFile() throws IOException, DukeDataCorruptedException {
+        Path filePath = Paths.get(this.filePath);
+
+        TaskList tasks;
+
+        // Check if the file exists before attempting to load
+        if (Files.exists(filePath)) {
+            try {
+                // Read tasks from the file
+                List<String> lines = Files.readAllLines(filePath);
+                tasks = new TaskList(TaskListDecoder.decode(lines));
+            } catch (FileNotFoundException e) {
+                // If the file is not found, throw a more specific exception
+                throw new FileNotFoundException("Data file not found");
+            } catch (IOException e) {
+                throw new IOException("Error reading tasks from the file.", e);
+            } catch (DukeDataCorruptedException e) {
+                // Rethrow other exceptions, including DukeDataCorruptedException
+                throw e;
+            }
+        } else {
+            tasks = new TaskList(); // Initialize with an empty list if the file doesn't exist yet
+        }
+
+        return tasks;
+    }
+
+    /**
+     * Ensures that the data file exists; if not, it creates the file.
+     *
+     * @throws IOException If an error occurs while creating the file.
+     */
+    public void ensureDataFileExists() throws IOException {
+        Path filePath = Paths.get(this.filePath);
+
+        // Check if the file exists or create it
+        if (!Files.exists(filePath)) {
+            // Check if the parent directory exists, create it if not
+            if (!Files.exists(filePath.getParent())) {
+                Files.createDirectories(filePath.getParent());
+            }
+
+            // Create the data file
+            Files.createFile(filePath);
+        }
+    }
+}
+
+/**
+ * The TaskListEncoder class encodes a list of tasks into a format suitable for saving to a file.
+ */
+class TaskListEncoder {
+    /**
+     * Encodes a list of tasks into a list of strings.
+     *
+     * @param tasks The list of tasks to encode.
+     * @return The encoded list of strings.
+     */
+    public static List<String> encode(List<Task> tasks) {
```

Comment:
I think the method name here can be more descriptive, to say encodeTaskToString or encodeToString so the client is clear how they are transforming the input